Improve RecordFromMediaSource resilience.

FerronN opened the pull request:
Hi,

I recently added LiveTV functionality to my Jellyfin installation. When recording programs, I noticed that my recordings were much shorter then then they should be. For example, if I the recording should be 1,5 hours, it only recorded 1 to 30 minutes, but I couldn't find any errors. Probably because the streams I watch aren't as stable as you want to be. Although you don't notice this when watching live (buffering in the Android TV client).

I noticed that the current code doesn't have any error handling if the stream for whatever reasons stops before the durationToken cancellation is requested. So, I updated the code to handle this. For me this works perfectly now.

I think the recording should always try to reconnect as long as the linkedCancellationToken cancellation hasn't been requested.

Please let me know if you have feedback on my implementation. 

**Changes**
- Added retry logic for restarting the stream for method `RecordFromMediaSource` in `EncodedRecorder.cs`.
- Added a bufferstream to smooth out minor hiccups (shows less drops then the retry logic) for `CopyToAsyncInternal` in `StreamHelper.cs`.

Commit:
Improve RecordFromMediaSource resilience.

Commit:
Fix ABI compliance for StreamHelper.cs

Bond-009 commented on MediaBrowser.Model/IO/IODefaults.cs:
```diff
@@ -21,5 +21,10 @@ public static class IODefaults
+        /// <summary>
+        /// The default buffer stream buffer size.
+        /// </summary>
+        public const int BufferStreamBufferSize = 131072;
```
The other defaults are copied from the .NET source code, while this one isn't.

FerronN replied:
Ah I didn't know that. Should I put it in a different constant class or put the value directly into the bufferstream constructor?

FerronN commented on src/Jellyfin.LiveTv/IO/StreamHelper.cs:
```diff
@@ -106,8 +131,9 @@ private static async Task<int> CopyToAsyncInternal(Stream source, Stream destina
         {
             int bytesRead;
             int totalBytesRead = 0;
+            using var bufferedStream = new BufferedStream(source, IODefaults.BufferStreamBufferSize);
```
I will change that

JPVenson commented on src/Jellyfin.LiveTv/IO/DirectRecorder.cs:
```diff
@@ -81,32 +81,63 @@ await _streamHelper.CopyToAsync(
+                }
+                catch (Exception ex)
+                {
+                    retry++;
+                    if (retry >= maxRetries)
+                    {
+                        _logger.LogError("Stream failed permanently after retries. {0}", ex.Message);
+                        return;
+                    }
```
I would rather see polly being used for resiliancy

FerronN commented:
@JPVenson I've changed it to use Polly. However, I moved the `BufferedStream` instantiation back to `CopyToAsyncInternal`.

I can't really explain it, but my streams are sometimes cut short when the `BufferedStream` instantiation is done in `CopyUntilCancelled` and passed to `CopyToAsyncInternal`.

The logs don't show any logs when this happens. The timer even fires when the recording should be stopped.
For example, I had a recording that should take 1,5 hours, but the duration of the recording was only 2 and once 8 minutes, before I moved the `BufferedStream` back.

I will validate the lasted commit in the coming few days to see if I experience any issues.
